fe1-web/src/features/witness/network/index.ts

```diff
@@ -1,18 +1,63 @@
-import { ActionType, MessageRegistry, ObjectType } from 'core/network/jsonrpc/messages';
+import {
+  ActionType,
+  AfterProcessingHandler,
+  ObjectType,
+  ProcessableMessage,
+} from 'core/network/jsonrpc/messages';
 
+import { WitnessConfiguration } from '../interface';
 import { WitnessMessage } from './messages';
+import { WitnessingType, getWitnessRegistryEntry } from './messages/WitnessRegistry';
 import { handleWitnessMessage } from './WitnessHandler';
+import { requestWitnessMessage } from './WitnessMessageApi';
+
+/**
+ * Is executed after a message has been successfully handled.
+ * It handles the passive witnessing for messages and prepares
+ * the application store for the act of manually witnessing
+ * other messages
+ */
+const afterMessageProcessingHandler =
+  (
+    enabled: WitnessConfiguration['enabled'],
+    /* isLaoWitness: WitnessConfiguration['isLaoWitness'] */
+  ): AfterProcessingHandler =>
+  (msg: ProcessableMessage) => {
+    const entry = getWitnessRegistryEntry(msg.messageData);
+
+    if (entry) {
+      // we have a wintessing entry for this message type
+      switch (entry.type) {
+        case WitnessingType.PASSIVE:
+          if (enabled) {
+            requestWitnessMessage(msg.channel, msg.message_id);
+          }
+          break;
+
+        case WitnessingType.ACTIVE:
+          break;
+
+        case WitnessingType.NO_WITNESSING:
+        default:
+          break;
+      }
+    }
+  };
 
 /**
  * Configures the network callbacks in a MessageRegistry.
  *
- * @param registry - The MessageRegistry where we want to add the mappings
+ * @param config - The witness feature configuration object
  */
-export function configureNetwork(registry: MessageRegistry) {
-  registry.add(
+export const configureNetwork = (config: WitnessConfiguration) => {
+  config.messageRegistry.add(
     ObjectType.MESSAGE,
     ActionType.WITNESS,
-    handleWitnessMessage,
+    handleWitnessMessage(config.getCurrentLao),
     WitnessMessage.fromJson,
   );
-}
+
+  config.messageRegistry.addAfterProcessingHandler(
+    afterMessageProcessingHandler(config.enabled /* config.isLaoWitness */),
+  );
```

Comment:
here we're hooking up to message registry / message processing directly.
It feels a bit ad-hoc, even though the solution was developed to be generic.

what if instead we hooked up to the store and watched messages being marked as processed (e.g. the `processMessages` event in the `MessageReducer`) ? (as we discussed today for the election key problem)

I see more code / more logic as a clear disadvantage of what I'm saying, but it'd also be purely centralized in the `witness` feature and a feature couldn't, on its own, break the message registry logic.

as a completely different approach, we could also envision the `processMessages` action to be wrapped in a thunk, so that pre-processors and post-processors can be added at that level (in a completely transparent way to the message registry).

Any thoughts ? The current implementation is fine anyhow, I'm just wondering what would look better😄

Reply:
Yes I think that is much nicer! I'll try to implement this using a store watcher now :)